utils/fixed_decimal/src/decimal.rs

```diff
@@ -697,6 +697,395 @@ impl FromStr for FixedDecimal {
+                if mag < 0 {
+                    // If the target magnitude was negative, make
+                    // sure we update the lower magnitude to match it
+                    decimal.lower_magnitude = mag;
+                } else if decimal.lower_magnitude < 0 {
+                    // If the target magnitude was positive,
+                    // we may have truncated digits off the right
+                    // side, reset the lower magnitude
+                    decimal.lower_magnitude = 0;
+                }
+            }
+            DoublePrecision::SignificantDigits(sig) => {
+                let sig = sig as i16;
+                if sig == 0 {
+                    return Err(Error::Limit);
+                }
+                let n_digits = decimal.digits.len() as i16;
+
+                if sig < n_digits {
+                    let round_by = (n_digits - sig) as i16;
+                    decimal.round_digits(round_by as u16);
+                    debug_assert!(decimal.digits.len() >= sig as usize);
+                    if decimal.digits.len() < sig as usize {
+                        // it's possible the magnitude increased by 1 due to overflow. round it again
+                        decimal.round_digits(1);
+                        // In case of overflow it cannot overflow again due to rounding
+                        debug_assert!(decimal.digits.len() == sig as usize);
+                    }
+                }
+                let target_lowest_magnitude = decimal.magnitude - sig as i16 + 1;
+                if target_lowest_magnitude <= 0 {
+                    // If the target magnitude was negative, make sure we update
+                    // the lower magnitude to match it
+                    decimal.lower_magnitude = target_lowest_magnitude;
+                } else if decimal.lower_magnitude < 0 {
+                    // If the target magnitude was positive,
+                    // we may have truncated digits off the right side
+                    // of the decimal point, reset the lower magnitude
+                    decimal.lower_magnitude = 0;
+                }
+            }
+        }
+        Ok(decimal)
+    }
+    /// Internal function for parsing directly from floats using ryū
+    fn new_from_f64_raw(float: f64) -> Result<Self, Error> {
+        if !float.is_finite() {
+            return Err(Error::Limit);
+        }
+        // note: this does not heap allocate
+        let mut buf = ryu::Buffer::new();
+        let formatted = buf.format_finite(float);
+        Self::from_str(formatted)
+    }
+
+    /// Internal function to round off `n` digits
+    /// from the right
+    ///
+    /// `self` must have at least `n` digits
+    ///
+    /// This may end up adding a digit to the left!
+    ///
+    /// This will not change the number of significant digits, it simply exists
+    /// to *round* them (and will typically reduce the size of `self.digits`)
+    fn round_digits(&mut self, n: u16) {
```

Comment:
Nit: The `round_digits` function is a greater scope than I think belongs in this PR.  It is a very important function that requires some time to review closely and make sure to get it right.  I would prefer if we would defer it to a future PR.